solver/src/paraswap_solver/api.rs

```diff
@@ -0,0 +1,420 @@
+#![allow(dead_code)]
+
+use std::convert::TryFrom;
+
+use anyhow::{Context, Result};
+use ethcontract::{H160, U256};
+use reqwest::{Client, RequestBuilder, Url};
+use serde::{Deserialize, Serialize};
+use serde_json::{Map, Value};
+
+use model::u256_decimal;
+use web3::types::Bytes;
+
+const BASE_URL: &str = "https://apiv4.paraswap.io";
+
+/// Mockable implementation of the API for unit test
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+pub trait ParaswapApi {
+    async fn price(&self, query: PriceQuery) -> Result<PriceResponse>;
+    async fn transaction(
+        &self,
+        query: TransactionBuilderQuery,
+    ) -> Result<TransactionBuilderResponse>;
+}
+
+struct DefaultParaswapApi {
+    client: Client,
+}
+
+#[async_trait::async_trait]
+impl ParaswapApi for DefaultParaswapApi {
+    async fn price(&self, query: PriceQuery) -> Result<PriceResponse> {
+        reqwest::get(query.into_url())
+            .await
+            .context("PriceQuery failed")?
+            .json::<Value>()
+            .await
+            .context("PriceQuery result parsing failed")
+            .and_then(PriceResponse::try_from)
+    }
+    async fn transaction(
+        &self,
+        query: TransactionBuilderQuery,
+    ) -> Result<TransactionBuilderResponse> {
+        query
+            .into_request(&self.client)
+            .send()
+            .await
+            .context("TransactionBuilderQuery failed")?
+            .json()
+            .await
+            .context("TransactionBuilderQuery result parsing failed")
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum Side {
+    Buy,
+    Sell,
+}
```

Comment:
Could we reuse our `OrderKind` enum from the model crate? Or are we trying to keep this entirely self-contained?

Reply:
They are not directly related, so I wanted to keep them separate (and not couple the Paraswap API with out core data mdoels).